fl4health/clients/fenda_client.py

```diff
@@ -1,12 +1,220 @@
+from fl4health.utils.losses import Losses, LossMeterType
+from fl4health.utils.metrics import Metric, MetricMeterType
 
 
 class FendaClient(BasicClient):
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        metric_meter_type: MetricMeterType = MetricMeterType.AVERAGE,
+        checkpointer: Optional[TorchCheckpointer] = None,
+        temperature: float = 0.5,
+        perFCL_loss_weights: List[float] = [0.0, 0.0],
+        cos_sim_loss_weight: float = 0.0,
+        contrastive_loss_weight: float = 0.0,
+    ) -> None:
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            metric_meter_type=metric_meter_type,
+            checkpointer=checkpointer,
+        )
+        self.perFCL_loss_weights = perFCL_loss_weights
+        self.cos_sim_loss_weight = cos_sim_loss_weight
+        self.contrastive_loss_weight = contrastive_loss_weight
+        self.cos_sim = torch.nn.CosineSimilarity(dim=-1)
+        self.ce_criterion = torch.nn.CrossEntropyLoss().to(self.device)
+        self.temperature: float = temperature
+
+        # Need to save previous local module, global module and aggregated global module at each communication round
+        # to compute contrastive loss.
+        self.old_local_module: torch.nn.Module
+        self.old_global_module: torch.nn.Module
+        self.aggregated_global_module: torch.nn.Module
+
     def get_parameter_exchanger(self, config: Config) -> ParameterExchanger:
         assert isinstance(self.model, FendaModel)
         return FixedLayerExchanger(self.model.layers_to_exchange())
+
+    def predict(self, input: torch.Tensor) -> Dict[str, torch.Tensor]:
+
+        preds = self.model(input)
+
+        preds["old_local_features"] = self.old_local_module.forward(input).view(len(preds["local_features"]), -1)
+        preds["old_global_features"] = self.old_global_module.forward(input).view(len(preds["global_features"]), -1)
+        if self.perFCL_loss_weights[0] != 0.0 or self.perFCL_loss_weights[1] != 0.0:
+            preds["aggregated_global_features"] = self.aggregated_global_module.forward(input).view(
+                len(preds["global_features"]), -1
+            )
+
+        if isinstance(preds, dict):
+            return preds
+        elif isinstance(preds, torch.Tensor):
+            return {"prediction": preds}
+        else:
+            raise ValueError("Model forward did not return a tensor or dictionary or tensors")
+
+    def set_parameters(self, parameters: NDArrays, config: Config) -> None:
+
+        # Save the parameters of the old model
+        assert isinstance(self.model.local_module, torch.nn.Module)
+        assert isinstance(self.model.global_module, torch.nn.Module)
```

Comment:
Rather than doing an assert on these submodules, could you rather assert that self.model is a FENDA model, making these two submodules proper nn.Modules?

Reply:
These assertion were needed for mypy because of `copy.deepcopy` in the next lines. Also as we only need previous global module and local module I don't make a copy of whole module which also has head module.

Reply:
Yes, I can see why you need to do some kind of typing assertion for mypy and that you're not copying the full fenda model. However, I was suggesting we just assert that self.model is a FendaModel type and mypy should be able to confirm that the submodules have the proper type. I tested this 
```        
assert isinstance(self.model, FendaModel)
self.old_local_module = copy.deepcopy(self.model.local_module)
self.old_global_module = copy.deepcopy(self.model.global_module)
```
and it seems like mypy was happy with that. It's a very minor code simplification.